File: cpp_tests/source/benchmarks/bench_cell_lists.cpp
#include <random>
#include <iostream>

#include "pele/neighbor_iterator.h"
#include "pele/lj_cut.h"
#include "pele/lbfgs.h"

using namespace pele;

int main()
{
    std::default_random_engine generator;
    std::uniform_real_distribution<double> distribution(0,1);

    size_t natoms = 1600;
    double rcut = 2.;
    double density = 1.2;
    double boxl = std::pow(natoms / density * (4/3 * M_PI), 1./3);
    std::cout << "box length " << boxl << std::endl;
    Array<double> boxvec(3, boxl);

    Array<double> x(3*natoms);
    for (size_t i = 0; i < x.size(); ++i) {
        x[i] = distribution(generator) * boxl;
    }

    double ncellx_scale = 1.;
    auto lj = std::make_shared<LJCutPeriodicCellLists<3> >(4., 4., rcut, boxvec, ncellx_scale);

    std::cout << "energy " << lj->get_energy(x) << "\n";

    LBFGS lbfgs(lj, x);
    lbfgs.set_max_iter(100000);
    lbfgs.set_iprint(100);
    lbfgs.run();

}

File: cpp_tests/source/benchmarks/bench_cell_lists.py
import numpy as np

from pele.potentials._lj_cpp import LJCutCellLists
from pele.optimize import lbfgs_cpp

x0_cpp = np.genfromtxt("coords")

np.random.seed(0)

x = x0_cpp.copy().ravel()
natoms = x.size / 3
density = 1.2
L = (natoms * (4./3*np.pi) / density)**(1./3)
print "box length", L
boxvec = np.array([L]*3)
rcut = 2.

pot = LJCutCellLists(boxvec=boxvec, rcut=rcut, ncellx_scale=1.)


#x = np.random.uniform(0, boxvec[0], natoms*3)


if False:
    res = lbfgs_cpp(x, pot, tol=100)
    x = res.coords
    print "coords"
    np.set_printoptions(threshold=np.nan, precision=16, linewidth=100)
    print repr(res.coords.reshape(-1,3))
    raise Exception("stopping early")


print "initial energy", pot.getEnergy(x)
lbfgs_cpp(x, pot, iprint=100)


#for i in xrange(1):
#    print "displacement", i
#    x += np.random.uniform(-.2, .2, x.size)
#    lbfgs_cpp(x, pot, iprint=50)
